Binary search over ranges

batiati opened the pull request:
This PR adds the ability to binary search over ranges, following the same semantics of the existing `binary_search_*` functions.

- Renames `binary_search_*_raw` to `binary_search_*_upsert_index`.

- Removes `config.verify` in favor of `constants.verify`, and removes the validation if the input array is sorted.

- Introduces `config.mode = .lower_bound | .upper_bound` to define the behavior of the binary search when the array contains duplicated entries.

- Adds tests for both `.lower_bound` and `.upper_bound` modes, including new test cases for duplicated matches.

- Clarify the comments.

The new functions `binary_search_*_range` and `binary_search_*_range_upsert_indexes` will be used by the Scan API. The point of adding it now is to make the review simpler and to reduce the amount of code to rebase later.

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -19,7 +19,7 @@ pub const Config = struct {
 ///
 /// Doesn't perform the extra key comparison to determine if the match is exact.
-pub fn binary_search_values_raw(
+pub fn binary_search_values_upsert_index(
```
While we are at it, could you clarify in the comment which index do we return if there are equal elements? (I wonder if we could/should assert that all elements are unique? Might be good for things like ids and such).

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -129,13 +129,148 @@ pub inline fn binary_search_keys(
+    );
+}
+
+pub inline fn binary_search_values_range_upsert_indexes(
```
Could you add a short doc comment here? In particular, looking just at the signature, the meaning of `key_min` / `key_max` is not immediately obvious to me.

matklad replied:
Ahhhh, this is not `equal_range`, we are looking for two keys at the same time. Still worth a short comment

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -129,13 +129,148 @@ pub inline fn binary_search_keys(
+    key_max: Key,
+    comptime config: Config,
+) BinarySearchRangeUpsertIndexes {
+    const start = binary_search_values_upsert_index(
```
`assert(key_min <= key_max)`

matklad replied:
Or maybe strict equality, to let the caller branch to non-range version?

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -129,13 +129,148 @@ pub inline fn binary_search_keys(
+        .end = start,
+    };
+
+    const end = binary_search_values_upsert_index(
```
Yeah, its unclear what's the semantics here if there are duplicates. If we are looking for  `(3, 5)`  in `0, 1, 3, 3, 3, 3, 4, 4, 5, 5, 5, 5, 6, 7`, what result do we expect? 

Seems that we should either assert that there are no dups, OR actually do the lower/upper bound to return the largest range (so, we should use slightly different algos for min and max bound)

batiati replied:
~Clarified in the comment that we don't expect duplicated entries, and included an `assert` for that when `constants.verify == true`.~

batiati replied:
Clarified the comment and added tests to assert that if we are looking for `(3, 5)` in `0, 1, 3, 3, 3, 3, 4, 4, 5, 5, 5, 5, 6, 7`, the expected result is the first occurrence of `3` (index 2) and the last occurrence of `5` (index 11)

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -339,3 +562,128 @@ test "binary search: random" {
     }
 }
+
+test "binary search: range" {
```
Let's write this in the same style as `binary search: explicit`. Instead of manually `try std.desting.expect` for each result, define a helper functions which takes both the input and the output, and asserts inside. 

Duplication of expectations in the tests makes future refactors painful, so I like to introduce a helper testing function (and it also good to write tests in the same style)

batiati committed:
Adds the ability to binary search over ranges

Follows the same semantics of the existing `binary_search_*` functions.

- Renames `binary_search_*_raw` to `binary_search_*_upsert_index`.

- Removes `config.verify` in favor of `constants.verify`.

- Introduces `config.mode = .lower_bound | .upper_bound` to define the behavior of the binary search when the array contains duplicated entries.

- Adds tests for both `.lower_bound` and `.upper_bound` modes, including new test cases for duplicated matches.

- Clarify the comments.

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -30,7 +37,7 @@ pub fn binary_search_values_raw(
     if (values.len == 0) return 0;
 
-    if (config.verify) {
+    if (constants.verify) {
```
Hm, not sure that using `constants.verify` is right here, because it changes big-O complexity from `log(N)` to `N`, so this'll probably slow down simulator and such too much. 

It makes sense that binary search has it's own toggle for full verification.

matklad replied:
All the other instancies of verify I think are fine though, as they preserve logarithmic behavior.

batiati replied:
You're right. Now it makes sense why running the fuzz tests was taking so long 😅.

The unit test itself is the only place setting this flag, so I'm not convinced if we should keep this assertion at all, unless we had plans to toggle this flag in a less intrusive way, not dependent on each call site.

matklad replied:
yeah, let's just remove it. This assert checks the caller, but if the caller are tests, then it doesn't really matter.

batiati committed:
Restore `verify_sorted` flag

batiati committed:
Remove sorted input validation
